packages/session/src/lib/session.ts

```diff
@@ -0,0 +1,175 @@
+
+export class Session {
+  /**
+   * Create a new cryptographically secure session ID.
+   * @returns A new session ID
+   */
+  static createId(): string {
+    return crypto.randomUUID()
+  }
+
+  constructor(id = Session.createId(), initialData?: SessionData) {
+    this.#id = id
+    this.#valueMap = new Map(initialData?.[0] ? Object.entries(initialData[0]) : [])
+    this.#flashMap = new Map(initialData?.[1] ? Object.entries(initialData[1]) : [])
+    this.#nextMap = new Map()
+  }
+
+  #id: string
+  readonly #valueMap: Map<string, unknown>
+  readonly #flashMap: Map<string, unknown>
+  readonly #nextMap: Map<string, unknown>
+  #destroyed = false
+  #dirty = false
+
+  /**
+   * The raw session data in a format suitable for storage.
+   * @returns The session data
+   */
+  get data(): SessionData {
+    return this.destroyed
+      ? [{}, {}]
+      : [Object.fromEntries(this.#valueMap), Object.fromEntries(this.#nextMap)]
+  }
+
+  /**
+   * Mark this session as destroyed.
+   *
+   * This prevents all further modifications to the session.
+   */
+  destroy(): void {
+    this.#destroyed = true
+  }
+
+  /**
+   * Whether this session has been destroyed.
+   */
+  get destroyed(): boolean {
+    return this.#destroyed
+  }
+
+  #checkDestroyed(): void {
+    if (this.destroyed) {
+      throw new Error('Session has been destroyed')
+    }
+  }
+
+  /**
+   * Whether this session has been modified since it was last saved.
+   */
+  get dirty(): boolean {
+    return this.#dirty
+  }
+
+  /**
+   * Set a value in the session that will be available only for the next request.
+   * @param key The key of the value to flash
+   * @param value The value to flash
+   */
+  flash(key: string, value: unknown): void {
+    this.#checkDestroyed()
+    this.#nextMap.set(key, value)
+    this.#dirty = true
+  }
+
+  /**
+   * Get a value from the session.
+   * @param key The key of the value to get
+   * @returns The value for the given key
+   */
+  get(key: string): unknown {
+    if (this.destroyed) return undefined
+    return this.#valueMap.get(key) ?? this.#flashMap.get(key)
+  }
+
+  /**
+   * Check if a value is stored for the given key.
+   * @param key The key to check
+   * @returns `true` if a value is stored for the given key, `false` otherwise
+   */
+  has(key: string): boolean {
+    if (this.destroyed) return false
+    return this.#valueMap.has(key) || this.#flashMap.has(key)
+  }
+
+  /**
+   * The unique identifier for this session.
+   */
+  get id(): string {
+    return this.#id
+  }
+
+  /**
+   * Regenerate the session ID while preserving the session data.
+   * This should be called after login or other privilege changes.
+   */
+  regenerateId(): void {
+    this.#checkDestroyed()
+    this.#id = Session.createId()
+    this.#dirty = true
+  }
+
+  /**
+   * The number of key/value pairs in the session.
+   */
+  get size(): number {
+    return this.#valueMap.size + this.#flashMap.size
+  }
```

Comment:
Did we end up using this for anything?

Reply:
Nah, I ended up adding a `context.sessionStarted` flag instead and issuing a warning if someone starts using `context.session` without a `session()` middleware in the stack since it's a no-op. I still think `session.size` is a nice-to-have though, so I left it in.